### PR TITLE
Devcontainers setup

### PR DESCRIPTION
**Pull request by richardhunghhw**
With dev containers, you won’t have to manually build the docker file each time to run the code. Run the code directly in the dev container or codespace.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,15 +10,37 @@ Sporkfish is a Python-based chess engine. Chess programming techniques, although
...
+2. Continue through all prompts using default options. 
+
+This will setup a Docker container to run in Github instead of Docker Desktop. All else being equal, but an active internet connection will be required for this, and this will use up your monthly allowance for Github codespace. 
+
+#### Using DevContainer without VSCode
+
+You do not need to install VSCode with this option.
 
-This generates an interactive bash shell for you to run the program in.
+#TODO
```
Comment (KYLChiu) [inline README.md:43]: Could we keep the original run method as well pls? I'm aware of people who might not be using vscode

File: scripts/container-postCreate.sh [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#/bin/bash
+
+# Run pip install
+pip install -r requirements.txt
```
Comment (KYLChiu) [inline scripts/container-postCreate.sh:4]: Just curious, why do we use a post-create script in stead of installing in the dockerfile?

I'm thinking WRT to caching, does this command get cached? AFAIK docker caches if we don't change anything above this command

File: README.md [hunk 3]
```diff
@@ -10,15 +10,37 @@ Sporkfish is a Python-based chess engine. Chess programming techniques, although
 
 ### Set up
```
Comment (KYLChiu) [inline README.md:11]: If you don't mind, shall we just move Setup into its own section now? Its getting heavy now, we have sub-sub-sub-sub sections now ;)

File: README.md [hunk 4]
```diff
@@ -10,15 +10,37 @@ Sporkfish is a Python-based chess engine. Chess programming techniques, although
...
+- [Docker Desktop](https://www.docker.com/products/docker-desktop/)
+
+#### Using DevContainer with VSCode 
+
+1. Make sure Docker Desktop is up and running.
+2. Open up the Command Palette and run `Dev Containers: Rebuild Container` 
+3. A warning might pop up saying `Not all host requirements in devcontainer.json are met by the Docker daemon.`, disregard this and continue. 
+4. The window should reload and you will see [Dev Container] in the URL bar as well as on the status bar bottom right of the window to indicate your setup is complete. 
+
+This will setup a Docker container to run in Docker Desktop with all the neceessary dependencies (python, pip, git, ...) with the correct versions.
```
Comment (KYLChiu) [inline README.md:26]: [Typo] Necessary

File: README.md [hunk 5]
```diff
@@ -10,15 +10,37 @@ Sporkfish is a Python-based chess engine. Chess programming techniques, although
 
 ### Set up
 
-To quickly set up the environment, you can use docker. From a terminal in the root directory:
+As a general case `Using DevContainer with VSCode` or `Using Github Codespace`
```
Comment (KYLChiu) [inline README.md:13]: Could we change this to: 

We can quickly setup the environment by utilizing DevContainer in VSCode, employing GitHub Codespaces or using naive docker (or somethin similar). Then give the 3 options?

File: README.md [hunk 6]
```diff
@@ -10,15 +10,37 @@ Sporkfish is a Python-based chess engine. Chess programming techniques, although
...
-```
+Install: 
+- [VSCode](https://code.visualstudio.com/download)
+- [Docker Desktop](https://www.docker.com/products/docker-desktop/)
+
+#### Using DevContainer with VSCode 
+
+1. Make sure Docker Desktop is up and running.
+2. Open up the Command Palette and run `Dev Containers: Rebuild Container` 
+3. A warning might pop up saying `Not all host requirements in devcontainer.json are met by the Docker daemon.`, disregard this and continue.
```
Comment (KYLChiu) [inline README.md:23]: Could we get rid of this if we get rid of of line 35-39 in the devcontainer.json file?

File: .devcontainer/devcontainer.json [hunk 7]
```diff
@@ -0,0 +1,40 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the
+// README at: https://github.com/devcontainers/templates/tree/main/src/typescript-node
+{
+    "name": "Python 3.11 Bookworm",
+    // Or use a Dockerfile or Docker Compose file. More info: https://containers.dev/guide/dockerfile
+    // "image": "mcr.microsoft.com/devcontainers/typescript-node:0-20",
+    "build": {
+        "dockerfile": "Dockerfile"
```
Comment (KYLChiu) [inline .devcontainer/devcontainer.json:8]: Curious, why don't we use the root level dockerfile and instead use the newly added one? If we remove the CMD "bash" in the root level dockerfile, will this work?

Comment (richardhunghhw) [reply .devcontainer/devcontainer.json:8]: The root dockerfile copies your local code into the container as part of the build. This is different to devcontainers which mounts the code into the container. Best to keep it seperate till I can confirm everything works great in devcontainers also.

File: scripts/container-postCreate.sh (hunk 2 above)
Comment (richardhunghhw) [reply scripts/container-postCreate.sh:4]: The docker container is not built with the code. The code is mounted in during container init, hence nothing to run pip install on.

As to why we dont include the requirement.txt into the docker, thats not the ideal practice with devcontainers. Different discussion if we are building for release.

Comment (KYLChiu) [reply scripts/container-postCreate.sh:4]: Need to read up on how this works, any best references you recommend?

**Review by KYLChiu — APPROVED** [review]

**Pull request merged**